### ISSUE TITLE
Golang: fails to communicate with Google logs unless HTTP/2 disabled

### ISSUE THREAD

**Issue by AGWA** [google/certificate-transparency#1136]
Under Go 1.6, attempts to retrieve the STH from any Google log fail, unless I disable HTTP/2 by setting `GODEBUG=http2client=0`.  I don't have trouble with any of the non-Google logs.

Test program:

```go
package main

import (
        "github.com/google/certificate-transparency/go/client"
        "fmt"
)

func try (url string) {
        c := client.New(url)
        _, err := c.GetSTH()
        if err != nil {
                fmt.Printf("%s: Error: %s\n", url, err)
        } else {
                fmt.Printf("%s: No error\n", url)
        }
}

func main () {
        try("https://ct.googleapis.com/pilot")
        try("https://ct.googleapis.com/aviator")
        try("https://ct.googleapis.com/rocketeer")
        try("https://ct.googleapis.com/submariner")
        try("https://ct1.digicert-ct.com/log")
}
```

Results under Go 1.5 OR Go 1.6 with `GODEBUG=http2client=0` set:
```
https://ct.googleapis.com/pilot: No error
https://ct.googleapis.com/aviator: No error
https://ct.googleapis.com/rocketeer: No error
https://ct.googleapis.com/submariner: No error
https://ct1.digicert-ct.com/log: No error
```

Results under Go 1.6:
```
https://ct.googleapis.com/pilot: Error: invalid character '<' looking for beginning of value
https://ct.googleapis.com/aviator: Error: invalid character '<' looking for beginning of value
https://ct.googleapis.com/rocketeer: Error: invalid character '<' looking for beginning of value
https://ct.googleapis.com/submariner: Error: invalid character '<' looking for beginning of value
https://ct1.digicert-ct.com/log: No error
```

Specifically, the Google server is returning the error `Error 400 (Bad Request)!!1` [sic].

**Comment by jmhodges** [comment]
This is probably a bug in the net/http library in Go 1.6 and should probably go over in the https://gihtub.com/golang/go issues.

**Comment by benlaurie** [comment]
https://github.com/google/certificate-transparency/blob/master/go/client/logclient.go#L127 appears to be the problem. Removing that line should work around it.

**Comment by AGWA** [comment]
Removing the Keep-Alive header fixed it for me - thanks!

**Comment by raullenchai** [comment]
Go 1.6 fix is still needed. Keep-Alive (and other Connection-Specific headers) should be ignored/removed if the underlying transport is HTTP2, thanks to https://http2.github.io/http2-spec/#rfc.section.8.1.2.2:
"An endpoint MUST NOT generate an HTTP/2 message containing connection-specific header fields; any message containing connection-specific header fields MUST be treated as malformed. ... Such intermediaries SHOULD also remove other connection-specific header fields, such as Keep-Alive, Proxy-Connection, Transfer-Encoding, and Upgrade..."

**Comment by jmhodges** [comment]
I made a ticket for the net/http2 fix: https://github.com/golang/go/issues/15085

**Comment by jmhodges** [comment]
This got fixed in x/net/http2 https://go-review.googlesource.com/#/c/21482/ and was merged in as https://github.com/golang/net/commit/024ed629fd292398cfd43c9678a5bf004f7defdc

**Comment by jsha** [comment]
So we can close this, right?

**Comment by jmhodges** [comment]
Wellllll, the problem is going to be around until whenever net/http gets this commit dropped in it. I don't know when it'll go out. Not sure if it'll be in 1.6.2 (the release coming soon after the 1.6.1 security fix). We might be waiting until 1.7 and making folks include and config http2 themselves with mysterious errors before they do so.

Could totally be wrong and it's in the plans for 1.6.2 and maybe in a few weeks this is all moot.

**Comment by jsha** [comment]
Ah, thanks. I was confused about the relationship between x/net/http2 and net/http. Looks like Boulder will just have to disable HTTP2 temporarily till the fix is part of a Go release.

**Comment by AGWA** [comment]
golang/go#15085 is tagged with the Go1.7 milestone, so I assume the fix is not going to make it into 1.6.2.  Thus I'm inclined to re-open this issue, though I'm not sure what the CT package should do.  Drop the Keep-Alive header? Disable HTTP/2?

**Comment by jsha** [comment]
I think the CT package should drop the Keep-Alive header. That way things continue to work, but less efficiently, in older versions, but get an efficiency improvement in newer versions of Go.
